### ISSUE TITLE
Hercules Hangs Using -d Parameter on Centos 8.3

### ISSUE THREAD

**Issue by marXtevens** [SDL-Hercules-390/hyperion#366]
I just built Hercules version 4.4.9999.0-SDL (4.4.9999.0) on CentOS Linux release 8.3.2011.

The -d parameter hangs? the emulator. See the log file, hercules4-d-f.vm370dev.marXtevens.strangled.net.2021-01-27.log, below.

Without the -d parameter it runs just fine. See the log file, hercules4-f.vm370dev.marXtevens.strangled.net.2021-01-27.log, below.

Any suggestions? Need any other files?

* [hercules4-d-f.vm370dev.marXtevens.strangled.net.2021-01-27.zip](https://github.com/SDL-Hercules-390/hyperion/files/5884126/hercules4-d-f.vm370dev.marXtevens.strangled.net.2021-01-27.zip)
* [hercules4-f.vm370dev.marXtevens.strangled.net.2021-01-27.zip](https://github.com/SDL-Hercules-390/hyperion/files/5884127/hercules4-f.vm370dev.marXtevens.strangled.net.2021-01-27.zip)

 ... Mark A. Stevens

**Comment by Fish-Git** [comment]
> The -d parameter hangs? the emulator.

> Without the -d parameter it runs just fine.

Patient: "Doctor! Doctor! It hurts when I do this!"
Doctor: "Then don't do that!"

&nbsp; &nbsp; `;-)`

Why are you trying to run Hercules in daemon mode?

I didn't write that part of Hercules so I am largely unfamiliar with it, but I seem to recall that daemon mode is used in order to run Hercules silently in the background wherein all stdin, stdout, and stderr traffic is controlled by another process (e.g. an external GUI for example). That is to say, all messages that would normally appear on your terminal (screen) are instead sent over a pipe to the controlling process (i.e. the GUI) which it is then free to display to the user in however a fashion/manner it wants via its own user interface, and all command _input_ used to control Hercules  is sent _TO_ Hercules from the external controlling process (i.e. the GUI), which it receives from the user via its own user interface.

As far as I know, Hercules's `--daemon` mode was not meant for normal use, which takes us back to my original question: Why are you trying to run Hercules in daemon mode? Do you have a legitimate need to run Hercules in that mode? Do you know what you are doing? Or are you just playing around? (Hacking?)

Basically, what is your question/problem? What is the incorrect behavior that you feel is broken and needs to be fixed?

**Comment by marXtevens** [comment]
But Doctor, it didn't used to hurt. :-D

>Why are you trying to run Hercules in daemon mode?

I became used to -d while using Hercules v3.x where I could then use the web browser as my GUI/TUI, and have my xterm/terminal session back for use, for finding, examining, and organizing "tapes" I had collected in preparation for loading them to my VM/370, or especially, the TK4- version of the MVS 3.8j system. 

>As far as I know, Hercules's --daemon mode was not meant for normal use, which takes us back to my original >question: Why are you trying to run Hercules in daemon mode? Do you have a legitimate need to run Hercules in >that mode? Do you know what you are doing? Or are you just playing around? (Hacking?)

I have been wanting to switch to version 4, but had not had the time. Work kept getting in my way. Now that I am retired, I'm beginning to catch up and have more fun, which includes hacking/playing around.

Is my need legitimate? I think so, but if it doesn't work, I'm not going to sweat it. Might try to figure out why, at some point, as I like a challenge, now and again.

Yes, I do know what I am doing, and have the holes in my feet from when I didn't.

I spent a fair amount of time getting this far (building a server to support KVMs, installing ooRexx and other pre-reqs). I've been a *ix System Admin for a long time, along with being a VM/SP->VM//ESA Systems Programmer long ago, ... and I can even spell Windows! :-D Not trying to be a prig, just letting you know I'm fairly competent.

I missed the --daemon parameter. I know ... read ALL the manuals first. ;-) I re-ran my setup, and hercules continues to run, and I can log in. I used VM/370, as it seems to be more tolerant of having the rug yanked out from under it.

I have attached the log file from that run. 
[hercules4-daemon-f.vm370dev.marXtevens.strangled.net.2021-01-28.zip](https://github.com/SDL-Hercules-390/hyperion/files/5889974/hercules4-daemon-f.vm370dev.marXtevens.strangled.net.2021-01-28.zip)

1. I see, but don't understand the problem with the threads.
2. Seems I have problems with ooRexx libraries. I'll look into this and fix it, to see if it is related to my problem.
3. I see no errors concerning port 8081, nor the directory assigned.

Thanks for your time.

**Comment by Fish-Git** [comment]
> 1. I see, but don't understand the problem with the threads.

The `HHC00109E set_thread_priority( 5 ) failed` and other similar error/warning messages (e.g. `HHC90020W 'hthread_setschedparam()' failed`) are caused by the fact that you are running Hercules as a regular user who does not have permission to set thread priorities (as evidenced by message `HHC00018W Hercules is NOT running in elevated mode`).

I am not a *ix person myself _(I'm a Windows weenie, but I can spell "Linux" and "Unix"!)_, but I believe you either need to start Hercules as root or else grant it the `CAP_SYS_NICE`(?) privilege/permission:

*  https://serverfault.com/questions/96742/give-pthread-setschedparam-permissions-to-non-root-user-linux

Or, another option, build Hercules _without_ "capabilities"(?) support via the `--disable-capabilities` configure option, so that it won't even _try_ setting internal thread priorities (though doing so is _not_ recommended; being able to set internal thread priorities is an important part of proper Hercules functioning).

> 2. Seems I have problems with ooRexx libraries. I'll look into this and fix it, to see if it is related to my problem.

It's likely _not_ related, but that's just a guess. As I recall the `librexxutil.so` shared library is an optional "extra" part of ooRexx that version 4.x may have support but which was dropped (I think) starting with v5.0, so message `HHC17531W` can probably be ignored. Rexx functionality _should_ still work.

Of course, if you don't really need integrated Rexx support in Hercules, then you should probably build it without it via the `--disable-object-rexx` and `--disable-regina-rexx` configure options.

> 3. I see no errors concerning port 8081, nor the directory assigned.

If Hercules's HTTP Server interface is not working it might be because you are not starting it. Show us your Hercules configuration file. Hercules's configuration file has changed quite a lot between v3.x and v4.x, with many statements being deprecated/removed, new ones introduced, and existing ones changed/enhanced. Please refer to our Configuration File web page and Release Notes web pages for more information:

  * [Hercules Version 4: Configuration File](https://sdl-hercules-390.github.io/html/hercconf.html)
  * [What's new in Hercules](https://sdl-hercules-390.github.io/html/hercnew.html)
  * [Hercules Release Notes and Known Issues](https://sdl-hercules-390.github.io/html/hercrnot.html)

The new [`HTTP`](https://sdl-hercules-390.github.io/html/hercconf.html#HTTPPORT) configuration file statements for example, replace the old v3.x `HTTPROOT` and `HTTPPORT` statements.

It also wouldn't hurt to take a peek at our many README web pages too:

  * [SDL Hercules 4.x (Hyperion) README](https://github.com/SDL-Hercules-390/hyperion/blob/master/README.md)

_(Note: the README web pages are best viewed using a browser that can properly display `.md` "markdown" files or that support a plugin/extension to accomplish the same thing. I personally use **[Markdown Viewer](https://chrome.google.com/webstore/detail/markdown-viewer/ckkdlimhmcjmikdlpkmbgfkaikojcbjk?hl=en)** for Chrome.)_

Hope that helps!

**Comment by marXtevens** [comment]
Yes, it did help. A lot!

Earlier, I had tripped over, and upgraded several of the old HTTP* commands, however it took a while before I  discovered the missing "command" for starting the web interface, in the ./hercules.cnf file:

> HTTP START

Adding this to my configuration, allows the web GUI to work, with and without the --daemon option. The --daemon option does not behave as it does in V3, but I can live with that. It was a nice, but not necessary.

I had removed both versions of Rexx from my configuration, to reduce problems. I have both included in my current build.

As would happen in the IBM world, I believe it is correct/safe to close this as DOC. Thank you.

 ... Mark A. Stevens

**Comment by marXtevens** [comment]
PostScript.
Followed the links on setcap/getcap and ended up doing the following.
```
[xmas@vm370dev THE-3.2]$ sudo setcap 'cap_sys_nice=eip' /usr/local/bin/hercules
[sudo] password for xmas: 
[xmas@vm370dev THE-3.2]$ getcap /usr/local/bin/hercules
/usr/local/bin/hercules = cap_sys_nice+eip
[xmas@vm370dev THE-3.2]$
```
Which allowed the following ...

```
[xmas@vm370dev VM370]$ hercules -f sixpackv4.conf
HHC00100I Thread id 00007fa8185c3740, prio 5, name 'impl_thread' started
HHC00100I Thread id 00007fa815d89700, prio 4, name 'logger_thread' started
HHC01413I Hercules version 4.4.9999.0-SDL (4.4.9999.0)
HHC01414I (C) Copyright 1999-2020 by Roger Bowler, Jan Jaeger, and others
HHC01417I ** The SoftDevLabs version of Hercules **
HHC01415I Build date: Jan 28 2021 at 20:38:56
```
Wahoo!
 ... Mark A. Stevens

**Comment by Fish-Git** [comment]
Mark: (@marXtevens)

FYI:

> <pre>
> HHC01413I Hercules version <b><i>4.4.9999.0-SDL (4.4.9999.0)</i></b>
> </pre>

It looks like you built your Hercules from the downloadable .zip file source code rather than from the source code of a clone of the repository. As a result, your reported version is an unhelpful non-specific _generic_ version string instead of a much more helpful _exact_ version string:

> <pre>
> HHC01413I Hercules version <b><i>4.4.9999.10396-SDL-ge21211ca (4.4.9999.10396)</i></b>
> </pre>

We would prefer that users who decide to build Hercules for themselves always do so from a `git clone` of the actual repository itself rather than from just the source code .zip file, as mentioned in **Step 1** of our "Installation and Operation" web page:

* [Hercules Version 4: Installation and Operation](https://sdl-hercules-390.github.io/html/hercinst.html#instnixmac)

To clone the source code repository, simply open a terminal window and enter the command:

```
git clone  https://github.com/SDL-Hercules-390/hyperion.git  <directory-name>
```

Then simply build Hercules normally. When you do that, you get the much more helpful _exact_ version string as shown above.

Knowing the _exact_ version that you are using is very important when it comes to reporting any problems you might have with Hercules. Without it, we can't know what fixes your copy of Hercules has and which ones it doesn't have. With an exact version string however, we can know _exactly_ what fixes (commits) your Hercules has and which ones it doesn't. The 8 hex characters following the "**-ge21211ca**" portion of the version string corresponds directly to which commit your Hercules was cloned from. E.g. "**e21211ca**" corresponds to commit e21211ca1792aaa827896abefe4fd1cc3f692293 as listed in the [repository commit log](https://github.com/SDL-Hercules-390/hyperion/commits/master).

Thanks.
